Tools/WinMLRunner/src/EventTraceHelper.cpp

```diff
@@ -0,0 +1,559 @@
+{ 
+    return TRUE; 
+}
+
+void EventTraceHelper::ProcessEventTrace(PTRACEHANDLE traceHandle)
+{
+    try
+    {
+        auto status = ProcessTrace(
+            traceHandle, 1, 0,
+            0); // this call blocks until either the session is stopped or an exception is occurred in event_callback
+        if (status != ERROR_SUCCESS && status != ERROR_CANCELLED)
+        {
+            printf("ProcessTrace failed with %lu\n", status);
+        }
+    }
+    catch (exception ex)
+    {
+        printf("Exception when processing event trace: %s\n", ex.what());
+        // catch exceptions occurred in event_callback
+    }
+}
+
+void EventTraceHelper::Start()
+{
+    int bufferSize = sizeof(EVENT_TRACE_PROPERTIES) + (sizeof(LOGSESSION_NAME) + 1) * sizeof(wchar_t);
+    sessionProperties = static_cast<PEVENT_TRACE_PROPERTIES>(malloc(bufferSize));
+    ZeroMemory(sessionProperties, bufferSize);
+
+    GUID guid;
+    UuidCreate(&guid);
+    sessionProperties->Wnode.BufferSize = static_cast<ULONG>(bufferSize);
+    sessionProperties->Wnode.Guid = guid;
+    sessionProperties->Wnode.ClientContext = 0;
+    sessionProperties->Wnode.Flags = WNODE_FLAG_TRACED_GUID;
+    sessionProperties->LogFileMode = EVENT_TRACE_REAL_TIME_MODE;
+    sessionProperties->LoggerNameOffset = sizeof(EVENT_TRACE_PROPERTIES);
+
+    auto hr = StartTrace(static_cast<PTRACEHANDLE>(&sessionHandle), LOGSESSION_NAME, sessionProperties);
+    if (hr != ERROR_SUCCESS)
+    {
+        printf("Error starting event trace: Trace already started %s\n", GetLastError());
+    }
+    auto status = EnableTraceEx2(sessionHandle, &DXGKRNL_PROVIDER_GUID, EVENT_CONTROL_CODE_ENABLE_PROVIDER,
+                                 TRACE_LEVEL_VERBOSE, 2, 0, 0, nullptr);
+
+    EVENT_TRACE_LOGFILE loggerInfo = {};
+
+    TRACE_LOGFILE_HEADER* pHeader = &loggerInfo.LogfileHeader;
+    ZeroMemory(&loggerInfo, sizeof(EVENT_TRACE_LOGFILE));
+    loggerInfo.LogFileMode = EVENT_TRACE_REAL_TIME_MODE;
+    loggerInfo.ProcessTraceMode = PROCESS_TRACE_MODE_REAL_TIME | PROCESS_TRACE_MODE_RAW_TIMESTAMP | PROCESS_TRACE_MODE_EVENT_RECORD;
+    loggerInfo.BufferCallback = BufferCallback;
+
+    // provide a callback whenever we get an event record
+    loggerInfo.EventRecordCallback = EventRecordCallback;
+    loggerInfo.Context = nullptr;
+    // LoggerName is the sessionName that we had provided in StartTrace
+    // For consuming events from ETL file we will provide path to ETL file.
+
+    loggerInfo.LoggerName = const_cast<LPWSTR>(LOGSESSION_NAME);
+
+    traceHandle = OpenTrace(&loggerInfo);
+    if (traceHandle == INVALID_PROCESSTRACE_HANDLE)
+    {
+        printf("Error opening event trace: OpenTrace failed with %lu\n", GetLastError());
+        throw std::runtime_error("Unable to open trace");
+    }
+
+    if (pHeader->PointerSize != sizeof(PVOID))
+        pHeader = (PTRACE_LOGFILE_HEADER)((PUCHAR)pHeader + 2 * (pHeader->PointerSize - sizeof(PVOID)));

```

Comment:
This seems to update pHeader, which is a pointer into loggerInfo, which is used to open the trace.

Why do we update this? From inspeaction it looks like this will do nothing?

Reply:
`pHeader` - ancient Hungarian... weird :b.

Reply:
Yeah will remove